### PR TITLE
Add custom result verifiers for min_by and max_by

### PR DESCRIPTION
**Pull request by kagamiori**
Summary:
Add custom result verifiers for min_by and max_by in aggregation fuzzer. These functions 
require custom result verifiers because they are order-sensitive while Presto doesn't 
honor order-by in aggregation funciton calls with them (https://github.com/prestodb/presto/issues/21431).

Differential Revision: D54865644

### CODE REVIEW

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                  {fmt::format("array_agg({}) as a", name_),
+                   "array_agg(expected) as e"})
+              .project({"remove_nulls(a) as a", "remove_nulls(e) as e"})
+              // Check that if actual result is not NULL, all elements in a[1]
+              // are also in e[1]. If the actual result is NULL, either the
+              // expected result is NULL too (when all rows in the group are
+              // masked out), or e[1] is empty (when all y in the group are
+              // NULL).
+              .project({fmt::format(
+                  "switch(cardinality(a) > 0, cardinality(e) > 0 and reduce(a[1], 0, (s, x) -> if (s < 0, -1, if (x is null, if (cardinality(e[1]) > cardinality(remove_nulls(e[1])), 1, -1), if (\"$internal$contains\"(e[1], x), 1, -1))), s -> s >= 0), cardinality(e) = 0 or cardinality(e[1]) = 0)",
```
Comment (kgpai) [inline velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:199]: Its pretty hard to parse this switch/reduce , wonder if theres a way to simplify (or if not a better explanation than the comment above).

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                  aggregateTypeSql_)})
```
Comment (kgpai) [inline velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:200]: Doesnt look like the type is used in the format ?

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+          BOOLEAN(), true, numGroups, result->pool());
+      velox::test::VectorMaker vectorMaker(result->pool());
+      auto expectedRow = vectorMaker.rowVector({expected});
+
+      return assertEqualResults({expectedRow}, {contains});
+    };
+    assertTrueResults(plan);
+
+    if (minMaxByN_) {
+      auto subsetSource =
```
Comment (kgpai) [inline velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:237]: This part unfortunately was not clear to me (on why we are generating for subsetSource).

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h (hunk 2 above)
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:200]: Good catch. I'll remove aggregateTypeSql_ here.

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    // that will be ignored by set_union later.
+    if (minMaxByN_) {
+      expectColumn = fmt::format(
+          "if ({} is null, cast(NULL as {}), expected) as expected",
+          yColumn,
+          aggregateTypeSql_);
+    } else {
+      expectColumn = fmt::format(
+          "if ({} is null, cast(NULL as {}[]), expected) as expected",
+          yColumn,
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:91]: Yes, we only care about top-level NULL of the yColumn.

File: velox/functions/prestosql/aggregates/ArrayAggAggregate.cpp [hunk 5]
```diff
@@ -267,4 +267,35 @@ void registerArrayAggAggregate(
...
+    bool overwrite) {
+  std::vector<std::shared_ptr<exec::AggregateFunctionSignature>> signatures{
+      exec::AggregateFunctionSignatureBuilder()
+          .typeVariable("E")
+          .returnType("array(E)")
+          .intermediateType("array(E)")
+          .argumentType("E")
+          .build()};
+
+  auto name = prefix + "$internal$array_agg";
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/functions/prestosql/aggregates/ArrayAggAggregate.cpp:282]: This is intended. We need an array_agg function that never ignores NULL. The built-in array_agg function cannot be used because it could ignore NULLs if the corresponding query config is set.

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h (hunk 3 above)
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:237]: subsetSource is just a PlanNode we're going to pass to localPartition together with actualSource. The goal of this localPartition is to combine the `subset_` and `result` together.

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    int64_t n = 1;
+    if (minMaxByN_) {
+      n = extractNValue(aggregate, input);
+    }
+
+    superSetPlan =
+        superSetPlan
+            .singleAggregation(groupingKeysAndY, {makeArrayAggCall(aggregate)})
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:119]: `frame` in the following .window() call contains an order-by on the yColumn.

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    // We'll first compute array_agg grouped by the original grouping keys plus
+    // y. We then apply set_union on these results associated with the min/max n
+    // y. We find the min/max n y by using the row_number window function over
+    // partitions by the original grouping keys ordered by y.
+    auto yColumn = extractYColumnName(aggregate);
+    auto groupingKeysAndY = groupingKeys;
+    groupingKeysAndY.push_back(yColumn);
+    auto frame = makeWindowFrame(groupingKeys, yColumn);
+
+    // projectColumnsBeforeUnion is used to clean up the result of array_agg for
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h]: Explained in the update. Thanks!

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+#include "velox/core/PlanNode.h"
+#include "velox/exec/fuzzer/ResultVerifier.h"
+#include "velox/exec/tests/utils/AssertQueryBuilder.h"
+#include "velox/exec/tests/utils/PlanBuilder.h"
+#include "velox/vector/ComplexVector.h"
+#include "velox/vector/tests/utils/VectorMaker.h"
+
+namespace facebook::velox::exec::test {
+
+// MinMaxByResultVerifier works by computing a superset and a subset of the
```
[the remark this answers is not in the document]
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:29]: Added. Thanks!

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h (hunk 1 above)
Comment (kagamiori) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h]: Extracted the common logic in the update. Thanks!

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h (hunk 3 above)
Comment (kgpai) [reply velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:350]: Yes I get that, but why should result be in both subset and superset ?

File: velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h [hunk 9]
```diff
@@ -0,0 +1,487 @@
...
+
+  bool supportsCompare() override {
+    return false;
+  }
+
+  bool supportsVerify() override {
+    return true;
+  }
+
+  void initialize(
```
Comment (Yuhta) [inline velox/functions/prestosql/fuzzer/MinMaxByResultVerifier.h:59]: Would be nice move some of the large function into cpp file to reduce compilation time

**Review by Yuhta — APPROVED** [review]
LGTM, please address Krishna's comments before merge as well
